code/modules/buildmode/buildmode.dm

```diff
@@ -0,0 +1,161 @@
+#define BM_SWITCHSTATE_NONE	0
+#define BM_SWITCHSTATE_MODE	1
+#define BM_SWITCHSTATE_DIR	2
+
+/datum/buildmode
+	var/build_dir = SOUTH
+	var/datum/buildmode_mode/mode
+	var/client/holder
+
+	var/li_cb
+	var/lo_cb
+
+	var/list/buttons
+	// SECTION UI
+
+	// Switching management
+	var/switch_state = BM_SWITCHSTATE_NONE
+	var/switch_width = 5
+	// modeswitch UI
+	var/obj/screen/buildmode/mode/modebutton
+	var/list/modeswitch_buttons = list()
+	// dirswitch UI
+	var/obj/screen/buildmode/bdir/dirbutton
+	var/list/dirswitch_buttons = list()
+
+/datum/buildmode/New(client/c)
+	mode = new /datum/buildmode_mode/basic(src)
+	holder = c
+	buttons = list()
+	li_cb = CALLBACK(src, .proc/post_login)
+	lo_cb = CALLBACK(src, .proc/post_logout)
+	holder.player_details.post_login_callbacks += li_cb
+	holder.player_details.post_logout_callbacks += lo_cb
+	holder.show_popup_menus = FALSE
+	create_buttons()
+	holder.screen += buttons
+	holder.click_intercept = src
+	mode.enter_mode(src)
+	
+/datum/buildmode/proc/quit()
+	mode.exit_mode(src)
+	holder.screen -= buttons
+	holder.click_intercept = null
+	holder.show_popup_menus = TRUE
+	qdel(src)
+
+/datum/buildmode/Destroy()
+	close_switchstates()
+	holder.player_details.post_login_callbacks -= li_cb
+	holder.player_details.post_logout_callbacks -= lo_cb
+	holder = null
+	QDEL_NULL(mode)
+	QDEL_LIST(modeswitch_buttons)
+	QDEL_LIST(dirswitch_buttons)
+	return ..()
+
+/datum/buildmode/proc/post_login()
+	// since these will get wiped upon login
+	holder.screen += buttons
+
+/datum/buildmode/proc/post_logout()
```

Comment:
@AnturK I created it as I added callbacks to the `player_details` datum, and did not actually find a use for it. I'll remove it now.

Reply:
Consequently, should I remove the entire `player_details` logout (not login) callback system? After this removal, nothing will utilize it.